ocaml/typing/uniqueness_analysis.ml

```diff
@@ -437,21 +448,122 @@ end = struct
           the analysis will error.
       *)
       let occ = Maybe_aliased.extract_occurrence l1 in
-      force_aliased_multiuse l0 m1 First;
+      force_aliased_multiuse l0 m1 First Seq;
       aliased occ Aliased.Forced
     | Aliased _, Aliased _ -> m0
     | Maybe_unique l, Aliased _ ->
-      force_aliased_multiuse l m1 First;
+      force_aliased_multiuse l m1 First Seq;
       m1
     | Aliased _, Maybe_unique l ->
-      force_aliased_multiuse l m0 Second;
+      force_aliased_multiuse l m0 Second Seq;
       m0
     | Maybe_unique l0, Maybe_unique l1 ->
-      force_aliased_multiuse l0 m1 First;
-      force_aliased_multiuse l1 m0 Second;
+      force_aliased_multiuse l0 m1 First Seq;
+      force_aliased_multiuse l1 m0 Second Seq;
       aliased (Maybe_unique.extract_occurrence l0) Aliased.Forced
 end
 
+module Tag : sig
+  (** This module collects the tags of allocations.
+      When we overwrite a tag, this module checks that the
+      tag is equal to the old tag. This is to ensure that
+      the tag never changes during overwrites. Changing the
+      tag during overwrites is not supported by the multicore GC. *)
+
+  type t =
+    | Unknown  (** default: we don't know the tag *)
+    | TagKnown of Longident.t loc
+        (** We have learned the old tag (eg. from a pattern match). *)
+    | OverwrittenAs of Longident.t loc
+        (** The tag was overwritten to be this. *)
+
+  type error =
+    | IncompatibleOverwrites of
+        { left_tag : Longident.t loc;
+          right_tag : Longident.t loc
+        }
+    | ChangedTag of
+        { old_tag : Longident.t loc option;
+          new_tag : Longident.t loc
+        }
+
+  exception Error of error
+
+  (** Sequential composition *)
+  val seq : t -> t -> t
+
+  (** Non-deterministic choice *)
+  val choose : t -> t -> t
+
+  (** Parallel composition *)
+  val par : t -> t -> t
+
+  (** We may not overwrite a tag we don't know.
+      At the end of the analysis, this function should
+      be called to ensure all overwrites are on known tags. *)
+  val check_no_remaining_overwritten_as : t -> unit
+end = struct
+  type t =
+    | Unknown
+    | TagKnown of Longident.t loc
```

Comment:
I still want the `Longident.t` for error messages, because `Types.tag` does not contain the user-written name. But I now also keep `Types.tag` and use that for the comparison between tags.